ntoskrnl/config/cmse.c

```diff
@@ -278,14 +279,26 @@ CmpSecurityMethod(IN PVOID ObjectBody,
     /* Acquire the hive lock */
     CmpLockRegistry();
 
-    /* Acquire the KCB lock */
+    /*
+     * Acquire the KCB lock but only if it has not been
+     * ever locked before (this is to avoid push lock
+     * recursion).
+     */
     if (OperationCode == QuerySecurityDescriptor)
     {
-        CmpAcquireKcbLockShared(Kcb);
+        if (!CmpIsKcbLocked(Kcb))
```

Comment:
This check implies that locking is optional. If so, just don't lock.
But I'm guessing it's not really optional and this is working around some bug. Could you elaborate?

Reply:
There is the risk that we can deadlock inside the `CmpSecurityMethod` function if it happens that the KCB of the object body already has a push lock held and somebody still wants to acquire it. The chance of a potential deadlock to occur is higher, for example, when we are holding a lock ourselves while doing a cache lookup during a parse procedure of the registry, or in the other words, somebody wants to query security information of a key object meanwhile we are holding the lock. On top of that, registry parsing is a very common occurrence further increasing such risk.

Why is this of a problem now but not before is because we are going to treat KCB locks seriously, since ReactOS misses some necessary locks here and there in the registry parser. And there's code that lives up on the fact that nobody was holding a lock. This is exactly what it happened to me with `CmpSecurityMethod`, when I was trying to debug the first deadlock I noticed that this function was called in `CmpDoCreateChild` with a KCB lock already held.

Now that I am discussing this, perhaps the correct fix would be is to remove the `CmpSecurityMethod` call in `CmpDoCreateChild`  and directly invoke `CmpAssignSecurityDescriptor`. It makes sense because `CmpSecurityMethod` is a construct used by OB when creating the key type object with `ObCreateObjectType` so nobody else should be calling `CmpSecurityMethod` but only OB when it sees fit. Also it seems like Windows doesn't use `CmpSecurityMethod` when creating a link node but rather it just calls `CmpAssignSecurityDescriptor`. Although you would still need to check if someone holds a lock against the KCB in `CmpSecurityMethod`, if the requested operation is to query info from a security descriptor that is.

Reply:
Yeah, removing direct calls to `CmpSecurityMethod` definitely makes sense to me.
If you wanted (and were able) to check "is the current thread already holding this lock" to avoid acquiring it recursively, that could make sense (though I would hope that there'd be more elegant solutions).
But the current check is: if anyone is already holding this lock, skip acquiring it -- which entirely negates the purpose of the lock, so it can't possibly be a correct fix.

Reply:
We have `CmpIsKcbLockedExclusive` which checks if the ownership of the lock is held by the calling thread, I don't see any other solutions that could see fit. If you have anything else that comes in your mind then suggestions are welcome.